docs/developer_overview.md

```diff
@@ -0,0 +1,511 @@
+};
+```
+
+To use warp-scope algorithms, warp size has to be known at compilation time (provided as a template parameter). It's possible to diverge from using the default warp size, in this case:
+
+- If `LOGICAL_WARP_THREADS` is a power of two - warp is partitioned into *sub*-warps, each reducing its data independently from other *sub*-warps. The terminology used in CUB: `32` threads are called hardware warp. Groups with less than `32` threads are called *logical* or *virtual* warp since it doesn't correspond directly to any hardware unit.
+- If `LOGICAL_WARP_THREADS` is **not** a power of two - there's no partitioning. That is, only the first logical warp executes algorithm.
+
+It's important to note that `LEGACY_PTX_ARCH` has been recently deprecated. This parameter used to affect specialization selection (see below). It was conflicting with the PTX dispatch refactoring and limited NVHPC support.
+
+#### Specialization
+
+CUB promises SOL performance when used with NVIDIA hardware. To achieve this, we have to take advantage of hardware-specific features. This is achieved by algorithm specializations. 
+
+```cpp
+using InternalWarpReduce = cub::detail::conditional_t<
+  IS_POW_OF_TWO,
+  WarpReduceShfl<T, LOGICAL_WARP_THREADS>,  /* shuffle-based implementation */
+  WarpReduceSmem<T, LOGICAL_WARP_THREADS>>; /* smem-based implementation */
+```
+
+Specializations provide different shared memory requirements, so the actual `_TempStorage` type is defined as:
+
+```cpp
+typedef typename InternalWarpReduce::TempStorage _TempStorage;
+```
+
+and algorithm implementation look like:
+
+
+```cpp
+__device__ __forceinline__ T Sum(T input, int valid_items) {
+  return InternalWarpReduce(temp_storage)
+      .Reduce(input, valid_items, cub::Sum());
+}
+```
+
+Due to `LEGACY_PTX_ARCH` issues described above, we can't specialize on the PTX version. `NV_IF_TARGET` shall be used by specializations instead:
+
+```cpp
+template <typename T, int LOGICAL_WARP_THREADS, int LEGACY_PTX_ARCH = 0>
+struct WarpReduceShfl 
+{
+   
+    
+template <typename ReductionOp>
+__device__ __forceinline__ T ReduceImpl(T input, int valid_items,
+                                        ReductionOp reduction_op)
+{
+  // ... base case (SM < 80) ...
+}
+
+template <class U = T>
+__device__ __forceinline__
+  typename std::enable_if<std::is_same<int, U>::value ||
+                          std::is_same<unsigned int, U>::value,
+                          T>::type
+    ReduceImpl(T input, 
+               int /* valid_items */, 
+               cub::Sum /* reduction_op */) 
+{
+  T output = input;
+
+  NV_IF_TARGET(NV_PROVIDES_SM_80,
+               (output = __reduce_add_sync(member_mask, input);),
+               (output = ReduceImpl<cub::Sum>(
+                    input, LOGICAL_WARP_THREADS, cub::Sum{});));
+
+  return output;
+}
+    
+    
+};
+```
+
+Specializations are stored in the `cub/warp/specializations` directory.
+
+#### Temporary storage usage
```

Comment:
Given that the specialization section refers to the temporary storage, I think the temporary storage section should be moved above.